### ISSUE TITLE
keyboard remains locked indefinitely

### ISSUE THREAD

**Issue by MareoRaft** [vitorgalvao/alfred-workflows#13]
Upon activation, the workflow successfully locks the keyboard, but never unlocks it.  Putting the computer in and out of sleep does not fix the issue.  Logging out does not fix the issue.  Restarting the computer does.

Computer:
MacBook Air 13", base model, serial number C17L4CJWFSV7, purchased summer of 2013.
OS:
10.9.5 Mavericks
Alfred version:
v2.6 (374)

Thanks!

**Comment by vitorgalvao** [comment]
Interesting. It seems that on the 2014 Mac Air it does not work, while on the 2013, this happens. Let’s do this in steps.

First, start by making sure you have the [latest version of Alfred](http://www.alfredapp.com/#download) (yours is outdated) and of [MachineClean](http://www.packal.org/workflow/machineclean).

Does the trackpad still work? If it does, it means you can shutdown the computer without having to do it forcibly, right?

What are the outputs of the following commands (you’ll need to [Open a terminal](http://youtu.be/OijYweDKhz4))?
```shell
kextstat | grep -i 'keyboard'
kextstat | grep -i 'multitouch'
```

**Comment by MareoRaft** [comment]
Before updating Alfred, the output was

    59    0 0xffffff7f81af6000 0x7000     0x7000     com.apple.driver.AppleHIDKeyboard (170.15) <28 4 3>
    57    2 0xffffff7f81adb000 0x17000    0x17000    com.apple.driver.AppleMultitouchDriver (245.13.1) <28 5 4 3 1>

After updating Alfred, the output was the same.

After running MachineClean (newest version), letting it default to 50 secs, output was

    57    2 0xffffff7f81adb000 0x17000    0x17000    com.apple.driver.AppleMultitouchDriver (245.13.1) <28 5 4 3 1>

And after 50 seconds, the output remained the same.

The trackpad still works and everything functions normally except for the keyboard.

**Comment by vitorgalvao** [comment]
To be clear, does the trackpad lock and unlock as expected, or is it unaffected?

I’m not sure how willing you are to try commands to narrow down the issue. At least the trackpad still works, so you don’t have to forcibly shutdown the computer, but it’s still a bit of a pain.

What’s the output of (it should lock the keyboard, but should also give some kind of warning).

```bash
keyboard=$(kextstat | grep -i 'keyboard' | head -1 | awk '{ print $6 }') && sudo kextunload -bundle-id "${keyboard}" ; sleep 3 ; echo "+ $(kextstat | grep -i 'keyboard' | head -1 | awk '{ print $6 }')" && sudo kextload -bundle-id "${keyboard}"
```

And if you can, open debugging mode (the little bug icon on the top right corner)
![](https://camo.githubusercontent.com/85a0a99d04d7ccda0cd57bab2e2e8e22a311a5f6/687474703a2f2f692e696d6775722e636f6d2f586d534a77436b2e706e67)
and run the workflow. What’s the output?

**Comment by MareoRaft** [comment]
The trackpad is unaffected.  It is always fully functional.

The output of above command was:
    +
and it locked my keyboard indefinitely.

When I trigger the workflow, I first get a message "osascript wants to make changes.  allow this?" and I have to type my password.  Then the workflow continues.  Here is the full log: (that is *All* messages, not just Errors and Warnings)
    Starting debug for 'MachineClean'
    
    [INFO: alfred.workflow.input.keyword] Processing output 'alfred.workflow.action.script' with arg ''
    [ERROR: alfred.workflow.action.script] 2015-05-16 18:52:22.270 terminal-notifier[444:507] [!] Unable     to post a notification for the current user (Matthew), as it has no running NotificationCenter instance.
2015-05-16 18:53:19.802 terminal-notifier[476:507] [!] Unable to post a notification for the current user (Matthew), as it has no running NotificationCenter instance.

So basically, nothing useful as far as I can tell.  (I disabled my notification center).

**Comment by vitorgalvao** [comment]
> does your workflow normally send a command to re-enable the keyboard?

Yes. What it does is unload the [kext](http://www.macbreaker.com/2012/01/what-are-kexts.html)s for the keyboard and trackpad, wait the chosen number of seconds, and loading them again. The output of that command, suggests what I was suspecting: in your case, unloading the kext for the keyboard is in someway making it inaccessible later for loading again. We might be able to fix that by using its full path.

What’s the output of

```bash
kextfind | grep -i 'keyboard'
kextfind | grep -i 'multitouch'
kextfind | grep -i 'trackpad'
```

And let’s try something different for the trackpad. Does `trackpad=$(kextstat | grep -i 'trackpad' | head -1 | awk '{ print $6 }') ; sudo kextunload -bundle-id "${trackpad}" ; sleep 5; sudo kextunload -bundle-id "${trackpad}"` work to lock the trackpad? It should lock it for 5 seconds, but naturally you should be prepared for it to not work again until a restart.

**Comment by ProveMath** [comment]
The output is:

    prove-math $ kextfind | grep -i 'keyboard'
    /System/Library/Extensions/AppleHIDKeyboard.kext
    /System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleBluetoothHIDKeyboard.kext    /System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleUSBBuiltinHIDKeyboard.kext    /System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleUSBHIDKeyboard.kext    /System/Library/Extensions/AppleUSBTopCase.kext/Contents/PlugIns/AppleUSBTCKeyboard.kext    prove-math $ kextfind | grep -i 'multitouch'    /System/Library/Extensions/AppleBluetoothMultitouch.kext    /System/Library/Extensions/AppleMultitouchDriver.kext    /System/Library/Extensions/AppleUSBMultitouch.kext    prove-math $ kextfind | grep -i 'trackpad'    /System/Library/Extensions/AppleUSBTopCase.kext/Contents/PlugIns/AppleUSBTrackpad.kext

**Comment by MareoRaft** [comment]
The output is:

    prove-math $ kextfind | grep -i 'keyboard'
    /System/Library/Extensions/AppleHIDKeyboard.kext
    /System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleBluetoothHIDKeyboard.kext
    /System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleUSBBuiltinHIDKeyboard.kext
    /System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleUSBHIDKeyboard.kext
    /System/Library/Extensions/AppleUSBTopCase.kext/Contents/PlugIns/AppleUSBTCKeyboard.kext
    prove-math $ kextfind | grep -i 'multitouch'
    /System/Library/Extensions/AppleBluetoothMultitouch.kext
    /System/Library/Extensions/AppleMultitouchDriver.kext
    /System/Library/Extensions/AppleUSBMultitouch.kext
    prove-math $ kextfind | grep -i 'trackpad'
    /System/Library/Extensions/AppleUSBTopCase.kext/Contents/PlugIns/AppleUSBTrackpad.kext

**Comment by vitorgalvao** [comment]
This looks good. We might be able to do something, at least for the keyboard (let’s focus on that for now).

```
keyboard='/System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleBluetoothHIDKeyboard.kext' ; sudo kextunload "${keyboard}" ; sleep 3 ; sudo kextload "${keyboard}"
keyboard='/System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleUSBBuiltinHIDKeyboard.kext' ; sudo kextunload "${keyboard}" ; sleep 3 ; sudo kextload "${keyboard}"
keyboard='/System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleUSBHIDKeyboard.kext' ; sudo kextunload "${keyboard}" ; sleep 3 ; sudo kextload "${keyboard}"
keyboard='/System/Library/Extensions/AppleUSBTopCase.kext/Contents/PlugIns/AppleUSBTCKeyboard.kext' ; sudo kextunload "${keyboard}" ; sleep 3 ; sudo kextload "${keyboard}"
```

Some of those lines should have no effect, while one of them should effectively lock your keyboard for 3 seconds, and succeed in unlocking it. Does it? Which one?

**Comment by MareoRaft** [comment]
Here are the outputs.  You can see that USBHID was different than the others (but it failed too, so it didn't lock the keyboard):

    ~ % keyboard='/System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleBluetoothHIDKeyboard.kext' ; sudo kextunload "${keyboard}" ; sleep 3 ; sudo kextload "${keyboard}"
    Password:
    (kernel) Kext com.apple.driver.AppleBluetoothHIDKeyboard not found for unload request.
    Failed to unload com.apple.driver.AppleBluetoothHIDKeyboard - (libkern/kext) not found.
    ~ %     keyboard='/System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleUSBBuiltinHIDKeyboard.kext' ; sudo kextunload "${keyboard}" ; sleep 3 ; sudo kextload "${keyboard}"
    (kernel) Kext com.apple.driver.AppleUSBBuiltinHIDKeyboard not found for unload request.
    Failed to unload com.apple.driver.AppleUSBBuiltinHIDKeyboard - (libkern/kext) not found.
    ~ % keyboard='/System/Library/Extensions/AppleHIDKeyboard.kext/Contents/PlugIns/AppleUSBHIDKeyboard.kext' ; sudo kextunload "${keyboard}" ; sleep 3 ; sudo kextload "${keyboard}"
    (kernel) Can't remove kext com.apple.driver.AppleUSBHIDKeyboard; services failed to terminate - 0xdc008018.
    Failed to unload com.apple.driver.AppleUSBHIDKeyboard - (libkern/kext) kext is in use or retained (cannot unload).
    ~ % keyboard='/System/Library/Extensions/AppleUSBTopCase.kext/Contents/PlugIns/AppleUSBTCKeyboard.kext' ; sudo kextunload "${keyboard}" ; sleep 3 ; sudo kextload "${keyboard}"
    (kernel) Kext com.apple.driver.AppleUSBTCKeyboard not found for unload request.
    Failed to unload com.apple.driver.AppleUSBTCKeyboard - (libkern/kext) not found.

**Comment by vitorgalvao** [comment]
Well, I’m out of ideas, really. Sorry I can’t do more, but diagnosing a machine’s model-specific issue without access to it over the span of weeks, one command at a time, is really hard.

There’s no good way to make a check for a machine’s model without accessing the internet, or I’d just block out Macbook Airs.

Throwing the towel on this one. Perhaps eventually there’ll be a pull request to fix this. Thank you for your cooperation on this.

**Comment by arthurnn** [comment]
interesing ,  i found this because i wanted to disable and enable back my internal keyboard but seems like only the disable command works and cannot disable it anymore, see:
```
[arthurnn@snowball Extensions]$ kextstat | grep -i 'keyboard' | head -1 | awk '{ print $6 }'
com.apple.driver.AppleHIDKeyboard
[arthurnn@snowball Extensions]$ sudo kextunload -bundle-id com.apple.driver.AppleHIDKeyboard
(kernel) Can't unload kext com.apple.driver.AppleHIDKeyboard; classes have instances:
(kernel)     Kext com.apple.driver.AppleHIDKeyboard class AppleHIDKeyboardEventDriverV2 has 1 instance.
Failed to unload com.apple.driver.AppleHIDKeyboard - (libkern/kext) kext is in use or retained (cannot unload).
[arthurnn@snowball Extensions]$ sudo kextload -bundle-id com.apple.driver.AppleHIDKeyboard
[arthurnn@snowball Extensions]$
```

i guess after the error of the unload command things crash and i cannot reload it.

any thoughts on that?

**Comment by vitorgalvao** [comment]
@arthurnn Yes, I had noticed that. It seems that after unloading it it simply vanishes, so it cannot be loaded again. Are you also on a Macbook Air?

**Comment by arthurnn** [comment]
yep macbook air 2015

**Comment by vitorgalvao** [comment]
[Can’t help you, then](https://github.com/vitorgalvao/alfred-workflows/issues/13#issuecomment-108563411). I’ve spent a lot of time on this already, with multiple people. I decided I won’t spend anymore on diagnosing. If someone has either:

+ A solution to the whole problem.
+ A solution to reliably identify Macbook Airs without accessing the internet.

I’ll take it. Otherwise, I won’t work on this specific problem any longer.

**Comment by vitorgalvao** [comment]
Found a solution to identify Macbook Airs without accessing the internet, so the workflow is now using that.
